instrumentation/google.golang.org/grpc/otelgrpc/test/interceptor_test.go

```diff
@@ -401,6 +435,78 @@ func TestStreamClientInterceptorOnBIDIStream(t *testing.T) {
 	_ = streamClient.CloseSend()
 }
 
+func TestStreamClientInterceptorEvents(t *testing.T) {
+	testCases := []struct {
+		Name     string
+		NumMsgs  int
+		NoEvents bool
+	}{
+		{Name: "With events", NumMsgs: 1, NoEvents: false},
+		{Name: "No events", NumMsgs: 1, NoEvents: true},
+		{Name: "10 messages with events", NumMsgs: 10, NoEvents: false},
+		{Name: "10 messages no events", NumMsgs: 10, NoEvents: true},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			defer goleak.VerifyNone(t)
+
+			method := "/github.com.serviceName/bar"
+			name := "github.com.serviceName/bar"
+			streamClient, sr := createInterceptedStreamClient(t, method, clientStreamOpts{NumRecvMsgs: testCase.NumMsgs, NoEvents: testCase.NoEvents})
+			_, ok := getSpanFromRecorder(sr, name)
+			require.False(t, ok, "span should not end while stream is open")
+
+			req := &grpc_testing.SimpleRequest{}
+			reply := &grpc_testing.SimpleResponse{}
+			var eventsAttr []map[attribute.Key]attribute.Value
+
+			// send and receive fake data
+			for i := 0; i < testCase.NumMsgs; i++ {
+				_ = streamClient.SendMsg(req)
+				_ = streamClient.RecvMsg(reply)
+				if !testCase.NoEvents {
+					eventsAttr = append(eventsAttr,
+						map[attribute.Key]attribute.Value{
+							otelgrpc.RPCMessageTypeKey: attribute.StringValue("SENT"),
+							otelgrpc.RPCMessageIDKey:   attribute.IntValue(i + 1),
+						},
+						map[attribute.Key]attribute.Value{
+							otelgrpc.RPCMessageTypeKey: attribute.StringValue("RECEIVED"),
+							otelgrpc.RPCMessageIDKey:   attribute.IntValue(i + 1),
+						},
+					)
+				}
+			}
+
+			// The stream has been exhausted so next read should get a EOF and the stream should be considered closed.
+			err := streamClient.RecvMsg(reply)
+			require.Equal(t, io.EOF, err)
+
+			// added retry because span end is called in separate go routine
+			var span trace.ReadOnlySpan
+			for retry := 0; retry < 5; retry++ {
+				span, ok = getSpanFromRecorder(sr, name)
+				if ok {
+					break
+				}
+				time.Sleep(time.Second * 1)
+			}
+			require.True(t, ok, "missing span %s", name)
```

Comment:
I think we could use https://pkg.go.dev/github.com/stretchr/testify/require?utm_source=godoc#Eventually or  https://pkg.go.dev/github.com/stretchr/testify/require?utm_source=godoc#EventuallyWithT instead

Reply:
Good idea.  I adjusted several places that used this logic to instead use `require.Eventually()`.